src/VisualStudio/Core/Def/LanguageService/AbstractPackage.cs

```diff
@@ -25,10 +25,34 @@ internal IComponentModel ComponentModel
         }
     }
 
-    protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
+    protected virtual void RegisterInitializationWork(PackageRegistrationTasks packageRegistrationTasks)
     {
-        _componentModel_doNotAccessDirectly = (IComponentModel?)await GetServiceAsync(typeof(SComponentModel)).ConfigureAwait(false);
-        Assumes.Present(_componentModel_doNotAccessDirectly);
+        // This treatment of registering work on the bg/main threads is a bit unique as we want the component model initialized at the beginning
+        // of whichever context is invoked first.
+        packageRegistrationTasks.AddTask(isMainThreadTask: false, task: EnsureComponentModelAsync);
+        packageRegistrationTasks.AddTask(isMainThreadTask: true, task: EnsureComponentModelAsync);
+
+        async Task EnsureComponentModelAsync(IProgress<ServiceProgressData> progress, PackageRegistrationTasks packageRegistrationTasks, CancellationToken token)
+        {
+            if (_componentModel_doNotAccessDirectly == null)
```

Comment:
might be worth doc'ing that since htis all executes serially, there is no threading concerns here.  and even if it executed concurrently, there shouldn't be concerns either since it's fine for either to 'win'.